File: client/app/leonardo.js
const req = require('request');
const fs = require('fs');

module.exports = {
    featureExtraction,
    similarityScoring
};

const LEONARDO_APIKEY = 'jGMB8R9KPK2MhNv7Tc9vTVGQ1mu7KLB0'
const LEONARDO_IMAGEFEATUREEXTRACTION_APIURL = 'https://sandbox.api.sap.com/ml/imagefeatureextraction/feature-extraction'
const LEONARDO_SIMILARITYSCORING_APIURL = 'https://sandbox.api.sap.com/ml/similarityscoring/similarity-scoring'

const headers = {
    'APIKey': LEONARDO_APIKEY,
    'Accept': 'application/json'
    // 'content-type' : 'multipart/form-data'
}

function featureExtraction(filename) {
    return new Promise((resolve, reject) => {
        req.post(LEONARDO_IMAGEFEATUREEXTRACTION_APIURL, {
            formData: {
                files: fs.createReadStream('./app/sample/' + filename)
            },
            json: true,
            headers: headers
        }, (err, res, body) => {
            if (err) { reject(err); }
            resolve(body);
        });
    });
}

function similarityScoring(vectors, numSimilarVectors = 1) {
    return new Promise((resolve, reject) => {
        req.post(LEONARDO_SIMILARITYSCORING_APIURL, {
            formData: {
                texts: JSON.stringify(vectors),
                options: JSON.stringify({ "numSimilarVectors": numSimilarVectors })
            },
            json: true,
            headers: headers
        }, (err, res, body) => {
            if (err) { reject(err); }
            resolve(body);
        });
    });
}

File: server/leonardo.py
# -*- coding: utf-8 -*-
import requests
import json

import settings

headers = {
    'APIKey': settings.LEONARDO_APIKEY,
    'Accept': 'application/json',
    # 'content-type' : 'multipart/form-data'
}


def feature_extraction(files):
    resp = requests.post(settings.LEONARDO_IMAGEFEATUREEXTRACTION_APIURL, files=files, headers=headers)
    if resp.status_code == 200:
        return json.loads(resp.text)['predictions']
    else:
        print('feature_extraction exception:', resp.status_code, resp.text)
        return []


def similarity_scoring(vectors, num_similar_vectors=1):
    data = {"texts": json.dumps(vectors),
            "options": json.dumps({"numSimilarVectors": num_similar_vectors})}
    resp = requests.post(settings.LEONARDO_SIMILARITYSCORING_APIURL, data=data, headers=headers)
    if resp.status_code == 200:
        return json.loads(resp.text)['predictions']
    else:
        print('similarity_scoring exception:', resp.status_code, resp.text)
        return []
